sql/mdl.cc

```diff
@@ -3438,7 +3438,14 @@ bool MDL_context::acquire_lock(MDL_request *mdl_request,
   /* There is a shared or exclusive lock on the object. */
   DEBUG_SYNC(get_thd(), "mdl_acquire_lock_wait");
 
-  find_deadlock();
+  /* We skip deadlock checks for ACL cache lock as there is severe contention
+  on internal RW lock of MDL lock. Since ACL cache lock is used during
+  connection process, all connections get stuck and will soon run out of
+  max_connections. All we need is good RW lock implementation. By removing this
+  deadlock checks, we remove contention on internal RW lock of MDL lock. This
+  means if ever there is deadlock due to code bug, we rely on
+  ACL_CACHE_LOCK_TIMEOUT value. */
+  if (mdl_request->key.mdl_namespace() != MDL_key::ACL_CACHE) find_deadlock();
```

Comment:
I'm fine with the current approach as it is perfect for now. 

Please consider this as a heads up that hat Oracle is implementing a deadlock detection infrastructure in replication using MDLs to avoid deadlocks with MDLs which may land anytime. Once it happens, we may have to re-evaluate/revert this change in very near future. 

If you are interested in knowing how ACL_CACHE can inlove in deadlock on slave, then please execute the below test.

```
➜  mysql-test git:(8.0) ✗ cat ../../mysql-test/suite/rpl/t/rpl_mts_spco_acl_deadlock.test
--source include/master-slave.inc

CREATE TABLE t1(a INT PRIMARY KEY);
CREATE TABLE t2(a INT PRIMARY KEY);

--source include/sync_slave_sql_with_master.inc

--source include/rpl_connection_slave.inc
BEGIN;
INSERT INTO t2 VALUES(1);

--source include/rpl_connection_master.inc
SET @@GLOBAL.DEBUG='+d,set_commit_parent_100';
INSERT INTO t2 VALUES(1); INSERT INTO t2 VALUES(2); CREATE USER u1;

--sleep 8
--source include/rpl_connection_slave.inc

SHOW PROCESSLIST;
SHOW PROCESSLIST;
SHOW PROCESSLIST;

--echo # we cant do anything more and slave server wont accept new
--echo # connections as well after this command

INSERT INTO t1 VALUES(1);
```
Run the test with 
```
./mtr rpl_mts_spco_acl_deadlock --mysqld=--slave-parallel-workers=8 --mysqld=--slave-parallel-type=1 --mysqld=--slave-preserve-commit-order=1 --mysqld=--binlog-format=statement --nocheck-testcases --mysqld=--innodb-lock-wait-timeout=5

```

Reply:
I was not able to reproduce this behaviour with or without fix..

on side note:
Lock Ordering issues (in code) are un-acceptable. We should acquire in same order. Server now has some lock ordering check mechanism. Unfortunately it is not automatic.

8.0.18 has one such fix to solve lock ordergin (deadlock) issue. Now all the ACL TABLES are opened first and only then the ACL cache lock.

Reply:
I am able to reproduce this hang. Agree that we have to evaluate if Oralce changes ACL MDL or ordering around that. Btw, even now, it is hang and not a deadlock because not all the parties involved are MDL locks.